src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -263,57 +266,95 @@ public static async Task<PackageSpec> GetProjectSpec(IDependencyGraphProject pro
 
             var projects = (await solutionManager.GetNuGetProjectsAsync()).OfType<IDependencyGraphProject>().ToList();
-            var knownProjects = projects.Select(e => e.MSBuildProjectPath).ToHashSet(PathUtility.GetStringComparerBasedOnOS());
+            var knownProjects = new ConcurrentDictionary<string, bool>(PathUtility.GetStringComparerBasedOnOS());
+            knownProjects.AddRange(projects.Select(e => e.MSBuildProjectPath)
+                .Select(proj => new KeyValuePair<string, bool>(proj, false)));
 
-            for (var i = 0; i < projects.Count; i++)
+            var options = new ExecutionDataflowBlockOptions()
             {
-                var (packageSpecs, projectAdditionalMessages) = await projects[i].GetPackageSpecsAndAdditionalMessagesAsync(context);
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+            var actionBlock = new ActionBlock<ProjectRestoreSpec>(GetProjectRestoreSpecAndAdditionalMessages, options);
 
-                if (projectAdditionalMessages != null && projectAdditionalMessages.Count > 0)
-                {
-                    if (allAdditionalMessages == null)
-                    {
-                        allAdditionalMessages = new List<IAssetsLogMessage>();
-                    }
+            foreach (IDependencyGraphProject project in projects)
+            {
+                var (packageSpecs, projectAdditionalMessages) = await project.GetPackageSpecsAndAdditionalMessagesAsync(context);
+                await actionBlock.SendAsync(new ProjectRestoreSpec(
+                                                dgSpec,
+                                                packageSpecs,
+                                                projectAdditionalMessages,
+                                                knownProjects,
+                                                allAdditionalMessages)
+                                            );
+            }
 
-                    allAdditionalMessages.AddRange(projectAdditionalMessages);
+            actionBlock.Complete();
+            await actionBlock.Completion;
+
+            // Return dg file
+            return (dgSpec, allAdditionalMessages.ToList());
+        }
+
+        private static void GetProjectRestoreSpecAndAdditionalMessages(
+            ProjectRestoreSpec restoreSpecData)
+        {
+            var dgSpec = restoreSpecData.DgSpec;
+
+            if (restoreSpecData.ProjectAdditionalMessages?.Any() ?? false)
+            {
+                foreach (var projectAdditionalMessage in restoreSpecData.ProjectAdditionalMessages)
+                {
+                    restoreSpecData.AllAdditionalMessages.Add(projectAdditionalMessage);
                 }
+            }
 
-                foreach (var packageSpec in packageSpecs)
+            foreach (var packageSpec in restoreSpecData.PackageSpecs)
+            {
+                lock (dgSpec)
```

Comment:
Just an idea, I don't know if anyone else thinks it's worthwhile doing, but we could avoid these locks and concurrent collections in the calling method, if `GetProjectRestoreSpecAndAdditionalMessages` returns a class or struct with all the data that needs to mutate, instead of mutating data in this method. If this method is a "pure function" (no mutation/side-effects), then we're gauranteed that running in parallel is as fast as possible. Then, `GetSolutionRestoreSpecAndAdditionalMessages` (the calling method) could modify the dgSpec, and other things, single-threaded.

I have no idea if it'll be faster or not. Honestly, I'm a bit surprised this implementation is much faster than doing it sequentially, because the work this method does looks quite simple, so I would expect that when run in parallel there's lots of lock contention.

Reply:
Good suggestion, I'll keep in mind for next time. Maybe not this time.